addons/binding/org.openhab.binding.loxone/src/main/java/org/openhab/binding/loxone/internal/core/LxControlFactory.java

```diff
@@ -0,0 +1,99 @@
+/**
+ * Copyright (c) 2010-2017 by the respective copyright holders.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.openhab.binding.loxone.internal.core;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.openhab.binding.loxone.internal.core.LxJsonApp3.LxJsonControl;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A factory of controls of Loxone Miniserver.
+ * It creates various types of control objects based on control type received from Miniserver.
+ *
+ * @author Pawel Pieczul
+ *
+ */
+class LxControlFactory {
+    static {
+        controls = new HashMap<>();
+        addType(LxControlDimmer.class);
+        addType(LxControlInfoOnlyAnalog.class);
+        addType(LxControlInfoOnlyDigital.class);
+        addType(LxControlJalousie.class);
+        addType(LxControlLightController.class);
+        addType(LxControlLightControllerV2.class);
+        addType(LxControlPushbutton.class);
+        addType(LxControlRadio.class);
+        addType(LxControlSwitch.class);
+        addType(LxControlTextState.class);
+        addType(LxControlTimedSwitch.class);
+    }
+
+    private static Map<String, Class<?>> controls;
+
+    /**
+     * Create a {@link LxControl} object for a control received from the Miniserver
+     *
+     * @param client
+     *            websocket client to facilitate communication with Miniserver
+     * @param uuid
+     *            UUID of the control to be created
+     * @param json
+     *            JSON describing the control as received from the Miniserver
+     * @param room
+     *            Room that this control belongs to
+     * @param category
+     *            Category that this control belongs to
+     * @return
+     *         created control object or null if error
+     */
+    static LxControl createControl(LxWsClient client, LxUuid uuid, LxJsonControl json, LxContainer room,
+            LxCategory category) {
+        if (json == null || json.type == null || json.name == null) {
+            return null;
+        }
+        String type = json.type.toLowerCase();
+        Class<?> c = controls.get(type);
+        if (c != null) {
+            try {
+                Constructor<?> constructor = c.getDeclaredConstructor(LxWsClient.class, LxUuid.class,
```

Comment:
Would it be hard to do it without reflection

Reply:
Can you point me somewhere with that? Previously I had just a long if else which does not seem like a esthetic nor effective way to do that. I could not really find a satisfying solution on the net in Java, other that using reflections. Need some help.

Reply:
If you properly isolate it I think that an old fashioned switch or if else can be a proper implementation of such factory. 

You could also have a map of factoring instances, which enable you to do:

controls.get(type).create(client, uuid, json, room, category);

The benefit is that you have way more compile time validations and less run-time surprises.

Reply:
@martinvw I tried to implement what you suggested, can you please see if I got it right. Thanks.